File: backend/src/types/marketplace.ts
// Marketplace IDL types
export type MarketplaceListing = {
  nftMint: string;
  seller: string;
  price: string;
  isActive: boolean;
};

// Cluster types
export type ClusterType = 'devnet' | 'testnet' | 'mainnet-beta';

// Placeholder IDL structure - replace with your actual IDL
export const MARKETPLACE_IDL = {
  "version": "0.1.0",
  "name": "marketplace",
  "instructions": [
    {
      "name": "createListing",
      "accounts": [
        { "name": "listing", "isMut": true, "isSigner": true },
        { "name": "marketplace", "isMut": false, "isSigner": false },
        { "name": "nftMint", "isMut": false, "isSigner": false },
        { "name": "seller", "isMut": true, "isSigner": true },
        { "name": "systemProgram", "isMut": false, "isSigner": false },
        { "name": "tokenProgram", "isMut": false, "isSigner": false }
      ],
      "args": [
        { "name": "price", "type": "u64" }
      ]
    },
    {
      "name": "buyNft",
      "accounts": [
        { "name": "listing", "isMut": true, "isSigner": false },
        { "name": "marketplace", "isMut": false, "isSigner": false },
        { "name": "nftMint", "isMut": false, "isSigner": false },
        { "name": "buyer", "isMut": true, "isSigner": true },
        { "name": "seller", "isMut": true, "isSigner": false },
        { "name": "systemProgram", "isMut": false, "isSigner": false },
        { "name": "tokenProgram", "isMut": false, "isSigner": false }
      ],
      "args": []
    },
    {
      "name": "cancelListing",
      "accounts": [
        { "name": "listing", "isMut": true, "isSigner": false },
        { "name": "marketplace", "isMut": false, "isSigner": false },
        { "name": "seller", "isMut": true, "isSigner": true },
        { "name": "systemProgram", "isMut": false, "isSigner": false }
      ],
      "args": []
    }
  ],
  "accounts": [
    {
      "name": "Listing",
      "type": {
        "kind": "struct",
        "fields": [
          { "name": "nftMint", "type": "publicKey" },
          { "name": "seller", "type": "publicKey" },
          { "name": "price", "type": "u64" },
          { "name": "isActive", "type": "bool" }
        ]
      }
    }
  ]
};

File: programs/marketplace/src/state/marketplace.rs
pub struct Listing {
    pub id: u64,
    pub seller: Pubkey,
    pub nft_mint: Pubkey,
    pub price: u64,
    pub is_active: bool,
}

pub struct Marketplace {
    pub listings: Vec<Listing>,
}

impl Marketplace {
    pub fn new() -> Self {
        Marketplace {
            listings: Vec::new(),
        }
    }

    pub fn add_listing(&mut self, listing: Listing) {
        self.listings.push(listing);
    }

    pub fn cancel_listing(&mut self, listing_id: u64) {
        if let Some(listing) = self.listings.iter_mut().find(|l| l.id == listing_id) {
            listing.is_active = false;
        }
    }

    pub fn buy_nft(&mut self, listing_id: u64) -> Option<&Listing> {
        if let Some(listing) = self.listings.iter_mut().find(|l| l.id == listing_id && l.is_active) {
            listing.is_active = false;
            return Some(listing);
        }
        None
    }
}